From: Christoph M. Becker <cmbecker69@gmx.de>
Date: Tue, 21 Jul 2020 11:07:43 +0200
Subject: Fix #79877: getimagesize function silently truncates after a null byte

We have to check for NUL bytes if `getimagesize()` has been called.

(cherry picked from commit ff577b04c0d250473a0ef46f8e332960fec3ca2c)

Origin: https://github.com/shivammathur/php-src-backports/commit/2c27d6bdd38bd8c4a27da5aded3477457e72c639
---
 ext/standard/image.c                   | 5 +++++
 ext/standard/tests/image/bug79877.phpt | 9 +++++++++
 2 files changed, 14 insertions(+)

diff --git a/ext/standard/image.c b/ext/standard/image.c
index 6c613382..94985717 100644
--- a/ext/standard/image.c
+++ b/ext/standard/image.c
@@ -1474,6 +1474,11 @@ static void php_getimagesize_from_any(INTERNAL_FUNCTION_PARAMETERS, int mode) {
 			return;
 	}
 
+	if (mode == FROM_PATH && CHECK_NULL_PATH(input, input_len)) {
+		php_error_docref(NULL, E_WARNING, "Invalid path");
+		return;
+	}
+
 	if (argc == 2) {
 		zval_dtor(info);
 		array_init(info);
diff --git a/ext/standard/tests/image/bug79877.phpt b/ext/standard/tests/image/bug79877.phpt
new file mode 100644
index 00000000..92e93e59
--- /dev/null
+++ b/ext/standard/tests/image/bug79877.phpt
@@ -0,0 +1,9 @@
+--TEST--
+Bug #79877 (getimagesize function silently truncates after a null byte)
+--FILE--
+<?php
+var_dump(getimagesize("/tmp/a.png\0xx"));
+?>
+--EXPECTF--
+Warning: getimagesize(): Invalid path in %s on line %d
+NULL
